inst/support.js

```diff
@@ -209,6 +234,109 @@ function numberIsInteger(x) {
 }
 
 
+// O(n^2) but does not use Set, which is not available in old v8
+function setDifference(a, b) {
+  var result = [];
+  for (var i = 0; i < a.length; i++) {
+    if (b.indexOf(a[i]) === -1) {
+      result.push(a[i]);
+    }
+  }
+  return result;
+}
+
+
+// nice behaviour both in and out of R
+function odinWarning(msg) {
+    try {
+        console.r.call("warning", msg)
+    } catch (e) {
+        console.warn(msg)
+    }
+}
+
+
+function odinMessage(msg) {
+    try {
+        console.r.call("message", msg)
+    } catch (e) {
+        console.warn(msg)
+    }
+}
+
+
+function flattenArray(value, name) {
+    var len = 1;
+    var dim = [];
+    var x = value;
+    while (Array.isArray(x)) {
+        dim.push(x.length);
+        len *= x.length;
+        x = x[0];
+    }
+    dim.reverse();
+
+    var data = flatten(value, []);
+
+    // Not a suffient check, but at least a necessary one:
+    if (len !== data.length) {
+        throw Error("Inconsistent array for '" + name + '"');
+    }
+
+    return {data: data, dim: dim};
+}
+
+
+// not all js versions have Array.prototype.flat?
```

Comment:
I'm not sure if Array.prototype.flat would flatten a 3D array in the same way either